File: sdk-go/standard_webhooks.go
package primitive

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"fmt"
	"strconv"
	"strings"
	"time"
)

// PrepareStandardWebhooksSecret strips the "whsec_" prefix if present,
// then base64-decodes the remainder to produce the raw HMAC key bytes.
func PrepareStandardWebhooksSecret(secret any) ([]byte, error) {
	var keyStr string
	switch v := secret.(type) {
	case string:
		keyStr = v
	case []byte:
		if len(v) == 0 {
			return nil, NewWebhookVerificationError(
				"MISSING_SECRET",
				"Webhook secret is required but was empty or not provided.",
				"",
			)
		}
		return v, nil
	case nil:
		return nil, NewWebhookVerificationError(
			"MISSING_SECRET",
			"Webhook secret is required but was empty or not provided.",
			"",
		)
	default:
		return nil, NewWebhookVerificationError(
			"MISSING_SECRET",
			fmt.Sprintf("Unsupported secret type: %T", secret),
			"",
		)
	}

	if keyStr == "" {
		return nil, NewWebhookVerificationError(
			"MISSING_SECRET",
			"Webhook secret is required but was empty or not provided.",
			"",
		)
	}

	if strings.HasPrefix(keyStr, whsecPrefix) {
		keyStr = keyStr[len(whsecPrefix):]
	}

	decoded, err := base64.StdEncoding.DecodeString(keyStr)
	if err != nil {
		// Try with padding
		decoded, err = base64.RawStdEncoding.DecodeString(keyStr)
		if err != nil {
			return nil, NewWebhookVerificationError(
				"MISSING_SECRET",
				"Standard Webhooks secret must be base64-encoded (optionally with whsec_ prefix).",
				"",
			)
		}
	}

	if len(decoded) == 0 {
		return nil, NewWebhookVerificationError(
			"MISSING_SECRET",
			"Webhook secret is required but was empty or not provided.",
			"",
		)
	}

	return decoded, nil
}

// parseStandardWebhooksSignatures parses the webhook-signature header
// into base64 signature strings. Only v1 entries are returned.
func parseStandardWebhooksSignatures(header string) []string {
	if header == "" {
		return nil
	}
	var signatures []string
	for _, entry := range strings.Split(header, " ") {
		entry = strings.TrimSpace(entry)
		if entry == "" {
			continue
		}
		idx := strings.Index(entry, ",")
		if idx == -1 {
			continue
		}
		version := entry[:idx]
		sig := entry[idx+1:]
		if version == "v1" && sig != "" {
			signatures = append(signatures, sig)
		}
	}
	return signatures
}

// SignStandardWebhooksPayload signs a payload using the Standard Webhooks format.
func SignStandardWebhooksPayload(rawBody any, secret any, msgID string, timestamps ...int64) (StandardWebhooksSignResult, error) {
	body, err := bodyToString(rawBody, "raw body")
	if err != nil {
		return StandardWebhooksSignResult{}, err
	}
	key, err := PrepareStandardWebhooksSecret(secret)
	if err != nil {
		return StandardWebhooksSignResult{}, err
	}
	timestamp := time.Now().Unix()
	if len(timestamps) > 0 {
		timestamp = timestamps[0]
	}
	mac := hmac.New(sha256.New, key)
	_, _ = mac.Write([]byte(fmt.Sprintf("%s.%d.%s", msgID, timestamp, body)))
	sig := base64.StdEncoding.EncodeToString(mac.Sum(nil))
	return StandardWebhooksSignResult{
		Signature: fmt.Sprintf("v1,%s", sig),
		MsgID:     msgID,
		Timestamp: timestamp,
	}, nil
}

// VerifyStandardWebhooksSignature verifies a Standard Webhooks signature.
func VerifyStandardWebhooksSignature(options StandardWebhooksVerifyOptions) (bool, error) {
	key, err := PrepareStandardWebhooksSecret(options.Secret)
	if err != nil {
		return false, err
	}
	if len(key) == 0 {
		return false, NewWebhookVerificationError(
			"MISSING_SECRET",
			"Webhook secret is required but was empty or not provided.",
			"",
		)
	}

	body, err := bodyToString(options.RawBody, "request body")
	if err != nil {
		return false, err
	}

	timestamp, err := strconv.ParseInt(options.Timestamp, 10, 64)
	if err != nil || timestamp < 0 {
		return false, NewWebhookVerificationError(
			"INVALID_SIGNATURE_HEADER",
			fmt.Sprintf("Invalid webhook-timestamp header: %q. Expected a unix timestamp in seconds.", options.Timestamp),
			"",
		)
	}

	now := time.Now().Unix()
	if options.NowSeconds != nil {
		now = *options.NowSeconds
	}
	tolerance := defaultToleranceSeconds
	if options.ToleranceSeconds != nil {
		tolerance = *options.ToleranceSeconds
	}
	age := now - timestamp
	if age > tolerance {
		return false, NewWebhookVerificationError(
			"TIMESTAMP_OUT_OF_RANGE",
			fmt.Sprintf("Webhook timestamp too old (%ds). Max age is %ds.", age, tolerance),
			"",
		)
	}
	if age < -futureToleranceSeconds {
		return false, NewWebhookVerificationError(
			"TIMESTAMP_OUT_OF_RANGE",
			"Webhook timestamp is too far in the future. Check server clock sync.",
			"",
		)
	}

	mac := hmac.New(sha256.New, key)
	_, _ = mac.Write([]byte(fmt.Sprintf("%s.%d.%s", options.MsgID, timestamp, body)))
	expectedSig := base64.StdEncoding.EncodeToString(mac.Sum(nil))

	signatures := parseStandardWebhooksSignatures(options.SignatureHeader)
	if len(signatures) == 0 {
		return false, NewWebhookVerificationError(
			"INVALID_SIGNATURE_HEADER",
			"Invalid webhook-signature header format. Expected: \"v1,<base64>\".",
			"",
		)
	}

	expectedBytes, _ := base64.StdEncoding.DecodeString(expectedSig)
	for _, sig := range signatures {
		sigBytes, err := base64.StdEncoding.DecodeString(sig)
		if err != nil {
			continue
		}
		if len(sigBytes) == len(expectedBytes) && hmac.Equal(sigBytes, expectedBytes) {
			return true, nil
		}
	}

	return false, NewWebhookVerificationError(
		"SIGNATURE_MISMATCH",
		"No valid signature found. Verify the webhook secret matches and you're using the raw request body (not re-serialized JSON).",
		"",
	)
}

// detectStandardWebhooksHeaders checks for Standard Webhooks headers.
// Returns all three values and ok=true if webhook-signature is present.
// If webhook-signature is found but other headers are missing, returns an error.
func detectStandardWebhooksHeaders(headers any) (msgID, timestamp, signature string, ok bool, err error) {
	signature = getHeaderValue(headers, StandardWebhookSignatureHeader)
	if !hasHeaderKey(headers, StandardWebhookSignatureHeader) {
		return "", "", "", false, nil
	}
	if signature == "" {
		return "", "", "", false, NewWebhookVerificationError(
			"INVALID_SIGNATURE_HEADER",
			"Empty webhook-signature header. Expected: \"v1,<base64>\".",
			"",
		)
	}
	msgID = getHeaderValue(headers, StandardWebhookIDHeader)
	timestamp = getHeaderValue(headers, StandardWebhookTimestampHeader)
	if msgID == "" || timestamp == "" {
		missing := "webhook-id"
		if msgID != "" {
			missing = "webhook-timestamp"
		}
		return "", "", "", false, NewWebhookVerificationError(
			"INVALID_SIGNATURE_HEADER",
			fmt.Sprintf("Found webhook-signature header but missing %s header. Standard Webhooks requires all three headers: webhook-id, webhook-timestamp, webhook-signature.", missing),
			"",
		)
	}
	return msgID, timestamp, signature, true, nil
}

File: sdk-node/src/webhook/standard-webhooks.ts
/**
 * Standard Webhooks Verification & Signing
 *
 * Implements the Standard Webhooks spec (standardwebhooks.com) for webhook
 * signature verification and payload signing.
 *
 * Header format:
 *   webhook-id: <msg_id>
 *   webhook-timestamp: <unix_seconds>
 *   webhook-signature: v1,<base64_hmac_sha256> [v1,<base64_2> ...]
 *
 * Signed payload format: "{msg_id}.{timestamp}.{raw_body}"
 *
 * Secrets are base64-encoded, optionally prefixed with "whsec_".
 */

import { createHmac, timingSafeEqual } from "node:crypto";
import { bufferToString } from "./encoding.js";
import { WebhookVerificationError } from "./errors.js";

const WHSEC_PREFIX = "whsec_";
const BASE64_PATTERN = /^[A-Za-z0-9+/]*={0,2}$/;

/** Default max age for webhook requests (5 minutes) */
const DEFAULT_TOLERANCE_SECONDS = 5 * 60;

/** Future clock skew tolerance (1 minute) */
const FUTURE_TOLERANCE_SECONDS = 60;

/** Standard Webhooks header names */
export const STANDARD_WEBHOOK_ID_HEADER = "webhook-id";
export const STANDARD_WEBHOOK_TIMESTAMP_HEADER = "webhook-timestamp";
export const STANDARD_WEBHOOK_SIGNATURE_HEADER = "webhook-signature";

/**
 * Result from signing a payload in Standard Webhooks format.
 */
export interface StandardWebhooksSignResult {
  /** The webhook-signature header value (e.g. "v1,<base64>") */
  signature: string;
  /** The message ID used for signing */
  msgId: string;
  /** Unix timestamp used for signing */
  timestamp: number;
}

/**
 * Options for verifying a Standard Webhooks signature.
 */
export interface StandardWebhooksVerifyOptions {
  /** The raw HTTP request body */
  rawBody: string | Buffer;
  /** The webhook-id header value */
  msgId: string;
  /** The webhook-timestamp header value (string, parsed to number internally) */
  timestamp: string;
  /** The webhook-signature header value */
  signatureHeader: string;
  /** Your webhook secret (with or without whsec_ prefix) */
  secret: string | Buffer;
  /** Max age in seconds (default: 300) */
  toleranceSeconds?: number;
  /** Override current time for testing (unix seconds) */
  nowSeconds?: number;
}

/**
 * Prepare a Standard Webhooks secret for HMAC computation.
 *
 * Strips the "whsec_" prefix if present, then base64-decodes the remainder
 * to produce the raw HMAC key bytes.
 *
 * @internal
 */
export function prepareStandardWebhooksSecret(secret: string | Buffer): Buffer {
  if (Buffer.isBuffer(secret)) {
    if (secret.length === 0) {
      throw new WebhookVerificationError(
        "MISSING_SECRET",
        "Webhook secret is required but was empty or not provided",
      );
    }
    return secret;
  }

  let keyStr = secret;
  if (keyStr.startsWith(WHSEC_PREFIX)) {
    keyStr = keyStr.slice(WHSEC_PREFIX.length);
  }

  if (!keyStr || !BASE64_PATTERN.test(keyStr)) {
    throw new WebhookVerificationError(
      "MISSING_SECRET",
      "Standard Webhooks secret must be base64-encoded (optionally with whsec_ prefix)",
    );
  }

  const decoded = Buffer.from(keyStr, "base64");
  if (decoded.length === 0) {
    throw new WebhookVerificationError(
      "MISSING_SECRET",
      "Webhook secret is required but was empty or not provided",
    );
  }
  return decoded;
}

/**
 * Parse the webhook-signature header into base64 signature strings.
 *
 * The header format is space-delimited entries like "v1,<base64>".
 * Only v1 entries are returned.
 *
 * @internal
 */
export function parseStandardWebhooksSignatures(header: string): string[] {
  if (!header) return [];

  const signatures: string[] = [];
  for (const entry of header.split(" ")) {
    const trimmed = entry.trim();
    if (!trimmed) continue;

    const commaIdx = trimmed.indexOf(",");
    if (commaIdx === -1) continue;

    const version = trimmed.slice(0, commaIdx);
    const sig = trimmed.slice(commaIdx + 1);

    if (version === "v1" && sig) {
      signatures.push(sig);
    }
  }

  return signatures;
}

/**
 * Sign a webhook payload using the Standard Webhooks format.
 *
 * @param rawBody - The raw JSON body string to sign
 * @param secret - The webhook secret (with or without whsec_ prefix)
 * @param msgId - The message ID (used in webhook-id header)
 * @param timestamp - Unix timestamp in seconds (defaults to current time)
 */
export function signStandardWebhooksPayload(
  rawBody: string | Buffer,
  secret: string | Buffer,
  msgId: string,
  timestamp?: number,
): StandardWebhooksSignResult {
  const ts = timestamp ?? Math.floor(Date.now() / 1000);
  const body =
    typeof rawBody === "string" ? rawBody : bufferToString(rawBody, "rawBody");
  const key = prepareStandardWebhooksSecret(secret);
  if (key.length === 0) {
    throw new WebhookVerificationError(
      "MISSING_SECRET",
      "Webhook secret is required but was empty or not provided",
    );
  }

  const signedPayload = `${msgId}.${ts}.${body}`;
  const sig = createHmac("sha256", key).update(signedPayload).digest("base64");

  return {
    signature: `v1,${sig}`,
    msgId,
    timestamp: ts,
  };
}

/**
 * Verify a Standard Webhooks signature.
 *
 * Throws `WebhookVerificationError` on failure with a specific error code.
 */
export function verifyStandardWebhooksSignature(
  opts: StandardWebhooksVerifyOptions,
): true {
  const {
    rawBody,
    msgId,
    timestamp: timestampStr,
    signatureHeader,
    secret,
    toleranceSeconds = DEFAULT_TOLERANCE_SECONDS,
    nowSeconds,
  } = opts;

  // Validate secret
  const key = prepareStandardWebhooksSecret(secret);
  if (key.length === 0) {
    throw new WebhookVerificationError(
      "MISSING_SECRET",
      "Webhook secret is required but was empty or not provided",
    );
  }

  // Parse timestamp
  if (!timestampStr || !/^\d+$/.test(timestampStr)) {
    throw new WebhookVerificationError(
      "INVALID_SIGNATURE_HEADER",
      `Invalid webhook-timestamp header: "${timestampStr}". Expected a unix timestamp in seconds`,
    );
  }
  const timestamp = Number(timestampStr);
  if (!Number.isInteger(timestamp) || timestamp < 0) {
    throw new WebhookVerificationError(
      "INVALID_SIGNATURE_HEADER",
      `Invalid webhook-timestamp header: "${timestampStr}". Expected a unix timestamp in seconds`,
    );
  }

  // Timestamp tolerance checks
  const now = nowSeconds ?? Math.floor(Date.now() / 1000);
  const age = now - timestamp;

  if (age > toleranceSeconds) {
    throw new WebhookVerificationError(
      "TIMESTAMP_OUT_OF_RANGE",
      `Webhook timestamp too old (${age}s). Max age is ${toleranceSeconds}s.`,
    );
  }

  if (age < -FUTURE_TOLERANCE_SECONDS) {
    throw new WebhookVerificationError(
      "TIMESTAMP_OUT_OF_RANGE",
      "Webhook timestamp is too far in the future. Check server clock sync.",
    );
  }

  // Compute expected signature
  const body =
    typeof rawBody === "string"
      ? rawBody
      : bufferToString(rawBody, "request body");

  const signedPayload = `${msgId}.${timestamp}.${body}`;
  const expectedSig = createHmac("sha256", key)
    .update(signedPayload)
    .digest("base64");

  // Parse and compare signatures
  const signatures = parseStandardWebhooksSignatures(signatureHeader);
  if (signatures.length === 0) {
    throw new WebhookVerificationError(
      "INVALID_SIGNATURE_HEADER",
      'Invalid webhook-signature header format. Expected: "v1,<base64>"',
    );
  }

  const expectedBytes = Buffer.from(expectedSig, "base64");

  for (const sig of signatures) {
    if (!BASE64_PATTERN.test(sig)) {
      continue;
    }
    const sigBytes = Buffer.from(sig, "base64");
    if (
      sigBytes.length === expectedBytes.length &&
      timingSafeEqual(sigBytes, expectedBytes)
    ) {
      return true;
    }
  }

  throw new WebhookVerificationError(
    "SIGNATURE_MISMATCH",
    "No valid signature found. Verify the webhook secret matches and you're using the raw request body (not re-serialized JSON).",
  );
}
